File: ios/Classes/RendererPlugin.swift
import Flutter
import UIKit

public class RendererPlugin: NSObject, FlutterPlugin {
    private var channel: FlutterMethodChannel?
    private var h265Decoder: H265Decoder?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "com.openup.streamline/renderer", binaryMessenger: registrar.messenger())
        let instance = RendererPlugin()
        instance.channel = channel
        instance.h265Decoder = H265Decoder(textureRegistry: registrar.textures())
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "init":
            guard let args = call.arguments as? [String: Any],
                  let width = args["width"] as? Int32,
                  let height = args["height"] as? Int32,
                  let vps = args["vps"] as? FlutterStandardTypedData,
                  let sps = args["sps"] as? FlutterStandardTypedData,
                  let pps = args["pps"] as? FlutterStandardTypedData  else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid initialization parameters", details: nil))
                return
            }
            let textureId = h265Decoder?.initialize(width: width, height: height, vps: vps.data, sps: sps.data, pps: pps.data)
            result(textureId)
            
        case "addH265Nal":
            guard let nal = call.arguments as? FlutterStandardTypedData else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid NAL data", details: nil))
                return
            }
            
            h265Decoder?.addH265Nal(nal.data)
            result(nil)
        case "dispose":
            h265Decoder?.cleanup();
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    public func detach(from engine: FlutterEngine) {
        h265Decoder?.cleanup()
        channel?.setMethodCallHandler(nil)
    }
}

File: android/src/main/kotlin/com/openup/renderer/RendererPlugin.kt
package com.openup.renderer

import android.content.Context
import androidx.annotation.NonNull

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class RendererPlugin: FlutterPlugin, MethodCallHandler {
    private lateinit var channel : MethodChannel
    private lateinit var h265Decoder : H265Decoder
    private var context: Context? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        h265Decoder = H265Decoder(flutterPluginBinding.getTextureRegistry())
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "com.openup.streamline/renderer")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "init" -> {
                val width = call.argument<Int>("width")!!
                val height = call.argument<Int>("height")!!
                val csd = call.argument<ByteArray>("csd")!!
                val textureId = h265Decoder.init(width, height, csd)
                result.success(textureId)
            }
            "addH265Nal" -> {
                val nal = call.arguments<ByteArray>()!!
                h265Decoder.addH265Nal(nal)
                result.success(null)
            }
            "dispose" -> {
              h265Decoder.cleanup()
              result.success(null)
            }
            "needsTransformation" -> {
              val needsTransformation = h265Decoder.needsTransformation();
              result.success(needsTransformation)
            }
            "sensorOrientation" -> {
                val localContext = context
                localContext?.let {
                    val sensorOrientation = h265Decoder.sensorOrientation(localContext);
                    result.success(sensorOrientation)
                } ?: run {
                    result.success(null)
                }
            }
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = null
        channel.setMethodCallHandler(null)
        h265Decoder.cleanup()
    }
}
